Porting index_add to structured kernels, add an out variant

krshrimali opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #65993

This PR attempts to port `index_add` to structured kernels, but does more than that:

* Adds an `out=` variant to `index_add`
* Revises `native_functions.yaml` registrations, to not have multiple entries and instead pass default value to `alpha`.
* Changes in `derivatives.yaml` file for autograd functioning
* Revises error messages, please see: https://github.com/pytorch/pytorch/pull/65993#issuecomment-945441615

Follow-up PRs in near future will attempt to refactor the OpInfo test, and will give another look at tests in `test/test_torch.py` for this function. (hence the use of ghstack for this)

~This is WIP because there are tests failing for `Dimname` variant on mobile/android builds, and I'm working on fixing them.~

Issue tracker: https://github.com/pytorch/pytorch/issues/55070

cc: @ezyang @bdhirsh @ysiraichi

Differential Revision: [D32646426](https://our.internmc.facebook.com/intern/diff/D32646426)

krshrimali committed:
Porting index_add to structured kernels, add an out variant

[ghstack-poisoned]

krshrimali committed:
Update on "Porting index_add to structured kernels, add an out variant"




[ghstack-poisoned]

krshrimali committed:
Update on "Porting index_add to structured kernels, add an out variant"


This PR attempts to port `index_add` to structured kernels, but does more than that:

* Adds an `out=` variant to `index_add`
* Revises `native_functions.yaml` registrations, to not have multiple entries and instead pass default value to `alpha`.
* Changes in `derivatives.yaml` file for autograd functioning

Follow-up PRs in near future will attempt to refactor the OpInfo test, and will give another look at tests in `test/test_torch.py` for this function. (hence the use of ghstack for this)

This is WIP because there are tests failing for `Dimname` variant on mobile/android builds, and I'm working on fixing them.

[ghstack-poisoned]

krshrimali committed:
Update on "Porting index_add to structured kernels, add an out variant"


This PR attempts to port `index_add` to structured kernels, but does more than that:

* Adds an `out=` variant to `index_add`
* Revises `native_functions.yaml` registrations, to not have multiple entries and instead pass default value to `alpha`.
* Changes in `derivatives.yaml` file for autograd functioning

Follow-up PRs in near future will attempt to refactor the OpInfo test, and will give another look at tests in `test/test_torch.py` for this function. (hence the use of ghstack for this)

This is WIP because there are tests failing for `Dimname` variant on mobile/android builds, and I'm working on fixing them.

[ghstack-poisoned]

krshrimali committed:
Update on "Porting index_add to structured kernels, add an out variant"


This PR attempts to port `index_add` to structured kernels, but does more than that:

* Adds an `out=` variant to `index_add`
* Revises `native_functions.yaml` registrations, to not have multiple entries and instead pass default value to `alpha`.
* Changes in `derivatives.yaml` file for autograd functioning

Follow-up PRs in near future will attempt to refactor the OpInfo test, and will give another look at tests in `test/test_torch.py` for this function. (hence the use of ghstack for this)

This is WIP because there are tests failing for `Dimname` variant on mobile/android builds, and I'm working on fixing them.

[ghstack-poisoned]

krshrimali commented:
This PR needs some attention in terms of discussions, and I'll welcome any opinions. Here are a few things that this PR introduces apart from porting to structured kernels:

1. Added an `out` variant for `index_add`.
2. The signatures (see below) were combined to form an _almost equivalent_ signature (please see below why I say _almost_).

```python
// explicitly passes alpha as 1
index_add(const Tensor& self, int64_t dim, const Tensor& index, const Tensor& source)
// takes alpha value as input
index_add.alpha(const Tensor& self, int64_t dim, const Tensor& index, const Tensor& source, *, Scalar alpha)
```

to:

```python
# default kwarg, when not passed - defaults to 1 (alpha)
index_add(const Tensor& self, int64_t dim, const Tensor& index, const Tensor& source, *, Scalar alpha=1)
```

(I may be wrong) - I feel that the new signature is better, as it reduces the amount of code for these variants. While porting to structured kernels, the new signature will help to make the code cleaner as well. And from the UX experience, I don't think it will break anything (I'm just not sure about the ONNX failure here, for XLA failure - please see below).

About the XLA Failure: (cc: @JackCaoG)

https://github.com/pytorch/xla/blob/e3e38164405efc37072dab07ea8e0c08a1072a26/torch_xla/csrc/tensor.h#L576-L580

- If we are happy/okay with adding an `out` variant for this function, we'll also need to add it to the XLA functions as well (happy to do this if everyone agrees that we need it).
- Currently, XLA doesn't support an `alpha` kwarg, which is interesting to me as it diverges from what we have in PyTorch core. Should we also think of adding an `alpha` kwarg in XLA?

Just a side note: the changes in XLA will also enable us to use the OpInfo for `index_add`. Currently, it's skipped if I'm not wrong.

I'm quite open to opinions and feedback here, as this is a design decision - I'm really looking forward to your thoughts. The current state of this PR will help act as a prototype to show how a probable implementation might look like.

cc: @ezyang @bdhirsh @JackCaoG @ysiraichi

JackCaoG commented:
Seems like pt/xla needs to adapt to a new `index_add` API and add support for `alpha`. I am happy to help but I would wait for this pytorch PR be approved from pytorch side first.

ysiraichi commented:
Hi @krshrimali. Good job porting this one! I will try to address some of your concerns, but we probably need more feedback to make a concrete decision.

> The signatures (see below) were combined to form an almost equivalent signature (please see below why I say almost).

Those look like good changes to me!

> And from the UX experience, I don't think it will break anything (I'm just not sure about the ONNX failure here, for XLA failure - please see below).

I think I agree with you, but I wouldn't trust myself too much on this one. I do know that dispatching functions relies on overloads and variants, so people that rely on those wouldn't be happy, I guess. Maybe we could use some more informed views! @ezyang  @bdhirsh What do you guys think?

> If we are happy/okay with adding an out variant for this function

I think that shouldn't be a problem.

bdhirsh commented:
>  I feel that the new signature is better, as it reduces the amount of code for these variants. While porting to structured kernels, the new signature will help to make the code cleaner as well. 

Agreed!

I found the PR that added the initial `alpha` overload of `index_add` here: https://github.com/pytorch/pytorch/pull/54176

There's a comment from @mruberry explicitly saying to add the new operator overload instead of adding a default argument to avoid breaking FC (I think), but that in the future we can remove the signature: https://github.com/pytorch/pytorch/pull/54176#discussion_r598839914. So this seems like a reasonable thing to do.

> I don't think it will break anything

Technically, it will break external backends that have registered kernels to either `index_add` or `index_add.alpha`. But this is generally a hard thing to avoid, and as you pointed out, this PR will require some accompanying PyTorch/XLA-side changes.

More specifically, PyTorch/XLA will need to add support for the alpha parameter. They actually won't need to do anything to support the `out=` variant - codegen will generate a kernel for them automatically :)

> Just a side note: the changes in XLA will also enable us to use the OpInfo for index_add. Currently, it's skipped if I'm not wrong.

Definitely support this!

krshrimali commented:
Thanks, @JackCaoG, @ysiraichi, and @bdhirsh for your valuable inputs on this one. Looks like we are all on the same page here. I'll send in the next commits to fix the failures and also coordinate with @JackCaoG to support the new API on the XLA side.

krshrimali committed:
Update on "Porting index_add to structured kernels, add an out variant"


This PR attempts to port `index_add` to structured kernels, but does more than that:

* Adds an `out=` variant to `index_add`
* Revises `native_functions.yaml` registrations, to not have multiple entries and instead pass default value to `alpha`.
* Changes in `derivatives.yaml` file for autograd functioning

Follow-up PRs in near future will attempt to refactor the OpInfo test, and will give another look at tests in `test/test_torch.py` for this function. (hence the use of ghstack for this)

~This is WIP because there are tests failing for `Dimname` variant on mobile/android builds, and I'm working on fixing them.~

Issue tracker: https://github.com/pytorch/pytorch/issues/55070

cc: ezyang bdhirsh ysiraichi 

[ghstack-poisoned]

ysiraichi reviewed (changes requested):
Hi @krshrimali. Nice work on this PR. I've left some comments there, so let me know if something is unclear.

bdhirsh commented:
@ysiraichi Just from trying out some examples on master, it looks to me like broadcasting fully works. e.g.
```
a = torch.ones(2, 3, 4, 5)
i  = torch.ones(7, dtype=torch.int)
a.index_add(0, i, torch.ones(7, 3, 4, 5))  # works, no broadcasting. The last argument matches `a` in all sizes except at index 0, where it matches `i.size()`.
a.index_add(0, i, torch.ones(7, 1, 4, 5))  # works, broadcasting!
a.index_add(0, i, torch.ones(7, 3, 1, 5))  # works, broadcasting!
a.index_add(0, i, torch.ones(7, 1, 1, 5))  # works, broadcasting!

# with dim=1 instead of 0
a.index_add(1, i, torch.ones(2, 7, 4, 5))  # works, no broadcasting
a.index_add(1, i, torch.ones(1, 7, 4, 5))  # works, broadcasting!
a.index_add(1, i, torch.ones(2, 7, 1, 5))  # works, broadcasting!
a.index_add(1, i, torch.ones(1, 7, 1, 5))  # works, broadcasting!
```

ysiraichi commented:
@bdhirsh https://github.com/pytorch/pytorch/pull/65993#issuecomment-944363938
Ah, I'm sorry. I mistyped that. It should have been:

```python
>>> torch.__version__
'1.9.0'
>>> a = torch.ones(5, 5, 5, 5)
>>> b = torch.ones(5, 5, 1, 5)
>>> i = torch.ones(5, dtype=torch.int)
>>> torch.index_add(b, 0, i, a) # this doesn't work.
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: output with shape [5, 1, 5] doesn't match the broadcast shape [5, 5, 5]
```

krshrimali committed:
Update on "Porting index_add to structured kernels, add an out variant"


This PR attempts to port `index_add` to structured kernels, but does more than that:

* Adds an `out=` variant to `index_add`
* Revises `native_functions.yaml` registrations, to not have multiple entries and instead pass default value to `alpha`.
* Changes in `derivatives.yaml` file for autograd functioning

Follow-up PRs in near future will attempt to refactor the OpInfo test, and will give another look at tests in `test/test_torch.py` for this function. (hence the use of ghstack for this)

~This is WIP because there are tests failing for `Dimname` variant on mobile/android builds, and I'm working on fixing them.~

Issue tracker: https://github.com/pytorch/pytorch/issues/55070

cc: ezyang bdhirsh ysiraichi 

[ghstack-poisoned]

krshrimali committed:
Update on "Porting index_add to structured kernels, add an out variant"


This PR attempts to port `index_add` to structured kernels, but does more than that:

* Adds an `out=` variant to `index_add`
* Revises `native_functions.yaml` registrations, to not have multiple entries and instead pass default value to `alpha`.
* Changes in `derivatives.yaml` file for autograd functioning

Follow-up PRs in near future will attempt to refactor the OpInfo test, and will give another look at tests in `test/test_torch.py` for this function. (hence the use of ghstack for this)

~This is WIP because there are tests failing for `Dimname` variant on mobile/android builds, and I'm working on fixing them.~

Issue tracker: https://github.com/pytorch/pytorch/issues/55070

cc: ezyang bdhirsh ysiraichi 

[ghstack-poisoned]

krshrimali committed:
Update on "Porting index_add to structured kernels, add an out variant"


This PR attempts to port `index_add` to structured kernels, but does more than that:

* Adds an `out=` variant to `index_add`
* Revises `native_functions.yaml` registrations, to not have multiple entries and instead pass default value to `alpha`.
* Changes in `derivatives.yaml` file for autograd functioning

Follow-up PRs in near future will attempt to refactor the OpInfo test, and will give another look at tests in `test/test_torch.py` for this function. (hence the use of ghstack for this)

~This is WIP because there are tests failing for `Dimname` variant on mobile/android builds, and I'm working on fixing them.~

Issue tracker: https://github.com/pytorch/pytorch/issues/55070

cc: ezyang bdhirsh ysiraichi 

[ghstack-poisoned]

krshrimali committed:
Update on "Porting index_add to structured kernels, add an out variant"


This PR attempts to port `index_add` to structured kernels, but does more than that:

* Adds an `out=` variant to `index_add`
* Revises `native_functions.yaml` registrations, to not have multiple entries and instead pass default value to `alpha`.
* Changes in `derivatives.yaml` file for autograd functioning

Follow-up PRs in near future will attempt to refactor the OpInfo test, and will give another look at tests in `test/test_torch.py` for this function. (hence the use of ghstack for this)

~This is WIP because there are tests failing for `Dimname` variant on mobile/android builds, and I'm working on fixing them.~

Issue tracker: https://github.com/pytorch/pytorch/issues/55070

cc: ezyang bdhirsh ysiraichi 

[ghstack-poisoned]

krshrimali committed:
Update on "Porting index_add to structured kernels, add an out variant"


This PR attempts to port `index_add` to structured kernels, but does more than that:

* Adds an `out=` variant to `index_add`
* Revises `native_functions.yaml` registrations, to not have multiple entries and instead pass default value to `alpha`.
* Changes in `derivatives.yaml` file for autograd functioning

Follow-up PRs in near future will attempt to refactor the OpInfo test, and will give another look at tests in `test/test_torch.py` for this function. (hence the use of ghstack for this)

~This is WIP because there are tests failing for `Dimname` variant on mobile/android builds, and I'm working on fixing them.~

Issue tracker: https://github.com/pytorch/pytorch/issues/55070

cc: ezyang bdhirsh ysiraichi 

[ghstack-poisoned]

krshrimali committed:
Update on "Porting index_add to structured kernels, add an out variant"


This PR attempts to port `index_add` to structured kernels, but does more than that:

* Adds an `out=` variant to `index_add`
* Revises `native_functions.yaml` registrations, to not have multiple entries and instead pass default value to `alpha`.
* Changes in `derivatives.yaml` file for autograd functioning

Follow-up PRs in near future will attempt to refactor the OpInfo test, and will give another look at tests in `test/test_torch.py` for this function. (hence the use of ghstack for this)

~This is WIP because there are tests failing for `Dimname` variant on mobile/android builds, and I'm working on fixing them.~

Issue tracker: https://github.com/pytorch/pytorch/issues/55070

cc: ezyang bdhirsh ysiraichi 

[ghstack-poisoned]

krshrimali committed:
Update on "Porting index_add to structured kernels, add an out variant"


This PR attempts to port `index_add` to structured kernels, but does more than that:

* Adds an `out=` variant to `index_add`
* Revises `native_functions.yaml` registrations, to not have multiple entries and instead pass default value to `alpha`.
* Changes in `derivatives.yaml` file for autograd functioning
* Revises error messages, please see: https://github.com/pytorch/pytorch/pull/65993#issuecomment-945441615

Follow-up PRs in near future will attempt to refactor the OpInfo test, and will give another look at tests in `test/test_torch.py` for this function. (hence the use of ghstack for this)

~This is WIP because there are tests failing for `Dimname` variant on mobile/android builds, and I'm working on fixing them.~

Issue tracker: https://github.com/pytorch/pytorch/issues/55070

cc: ezyang bdhirsh ysiraichi 

[ghstack-poisoned]

krshrimali committed:
Update on "Porting index_add to structured kernels, add an out variant"


This PR attempts to port `index_add` to structured kernels, but does more than that:

* Adds an `out=` variant to `index_add`
* Revises `native_functions.yaml` registrations, to not have multiple entries and instead pass default value to `alpha`.
* Changes in `derivatives.yaml` file for autograd functioning
* Revises error messages, please see: https://github.com/pytorch/pytorch/pull/65993#issuecomment-945441615

Follow-up PRs in near future will attempt to refactor the OpInfo test, and will give another look at tests in `test/test_torch.py` for this function. (hence the use of ghstack for this)

~This is WIP because there are tests failing for `Dimname` variant on mobile/android builds, and I'm working on fixing them.~

Issue tracker: https://github.com/pytorch/pytorch/issues/55070

cc: ezyang bdhirsh ysiraichi 

[ghstack-poisoned]

krshrimali committed:
Update on "Porting index_add to structured kernels, add an out variant"


This PR attempts to port `index_add` to structured kernels, but does more than that:

* Adds an `out=` variant to `index_add`
* Revises `native_functions.yaml` registrations, to not have multiple entries and instead pass default value to `alpha`.
* Changes in `derivatives.yaml` file for autograd functioning
* Revises error messages, please see: https://github.com/pytorch/pytorch/pull/65993#issuecomment-945441615

Follow-up PRs in near future will attempt to refactor the OpInfo test, and will give another look at tests in `test/test_torch.py` for this function. (hence the use of ghstack for this)

~This is WIP because there are tests failing for `Dimname` variant on mobile/android builds, and I'm working on fixing them.~

Issue tracker: https://github.com/pytorch/pytorch/issues/55070

cc: ezyang bdhirsh ysiraichi 

[ghstack-poisoned]

krshrimali committed:
Update on "Porting index_add to structured kernels, add an out variant"


This PR attempts to port `index_add` to structured kernels, but does more than that:

* Adds an `out=` variant to `index_add`
* Revises `native_functions.yaml` registrations, to not have multiple entries and instead pass default value to `alpha`.
* Changes in `derivatives.yaml` file for autograd functioning
* Revises error messages, please see: https://github.com/pytorch/pytorch/pull/65993#issuecomment-945441615

Follow-up PRs in near future will attempt to refactor the OpInfo test, and will give another look at tests in `test/test_torch.py` for this function. (hence the use of ghstack for this)

~This is WIP because there are tests failing for `Dimname` variant on mobile/android builds, and I'm working on fixing them.~

Issue tracker: https://github.com/pytorch/pytorch/issues/55070

cc: ezyang bdhirsh ysiraichi 

[ghstack-poisoned]

krshrimali committed:
Update on "Porting index_add to structured kernels, add an out variant"


This PR attempts to port `index_add` to structured kernels, but does more than that:

* Adds an `out=` variant to `index_add`
* Revises `native_functions.yaml` registrations, to not have multiple entries and instead pass default value to `alpha`.
* Changes in `derivatives.yaml` file for autograd functioning
* Revises error messages, please see: https://github.com/pytorch/pytorch/pull/65993#issuecomment-945441615

Follow-up PRs in near future will attempt to refactor the OpInfo test, and will give another look at tests in `test/test_torch.py` for this function. (hence the use of ghstack for this)

~This is WIP because there are tests failing for `Dimname` variant on mobile/android builds, and I'm working on fixing them.~

Issue tracker: https://github.com/pytorch/pytorch/issues/55070

cc: ezyang bdhirsh ysiraichi 

[ghstack-poisoned]

krshrimali committed:
Update on "Porting index_add to structured kernels, add an out variant"


This PR attempts to port `index_add` to structured kernels, but does more than that:

* Adds an `out=` variant to `index_add`
* Revises `native_functions.yaml` registrations, to not have multiple entries and instead pass default value to `alpha`.
* Changes in `derivatives.yaml` file for autograd functioning
* Revises error messages, please see: https://github.com/pytorch/pytorch/pull/65993#issuecomment-945441615

Follow-up PRs in near future will attempt to refactor the OpInfo test, and will give another look at tests in `test/test_torch.py` for this function. (hence the use of ghstack for this)

~This is WIP because there are tests failing for `Dimname` variant on mobile/android builds, and I'm working on fixing them.~

Issue tracker: https://github.com/pytorch/pytorch/issues/55070

cc: ezyang bdhirsh ysiraichi 

[ghstack-poisoned]

bdhirsh reviewed (approved):
@krshrimali ok, I think the last thing you need to change is to add a line to the [backwards_compatibility test here](https://github.com/pytorch/pytorch/blob/eea20bfa1592de494272dfaae0a357ff5139928d/test/backward_compatibility/check_backward_compatibility.py#L60). I think we're agreed that this isn't actually BC-breaking, but the test sees a change to `index_add.alpha` and `index_add_.alpha` (because... we're deleting them). You just need to add a line at the bottom of that list of each operator. For the date, maybe put the end of the year :) since if the date passes before this PR merges into master, the test will start failing again.

@JackCaoG I think this PR is ready for xla-side changes. Do you mind making a patch for index_add? (adding a defaultable `alpha` kwarg at the end).

krshrimali committed:
Update on "Porting index_add to structured kernels, add an out variant"


This PR attempts to port `index_add` to structured kernels, but does more than that:

* Adds an `out=` variant to `index_add`
* Revises `native_functions.yaml` registrations, to not have multiple entries and instead pass default value to `alpha`.
* Changes in `derivatives.yaml` file for autograd functioning
* Revises error messages, please see: https://github.com/pytorch/pytorch/pull/65993#issuecomment-945441615

Follow-up PRs in near future will attempt to refactor the OpInfo test, and will give another look at tests in `test/test_torch.py` for this function. (hence the use of ghstack for this)

~This is WIP because there are tests failing for `Dimname` variant on mobile/android builds, and I'm working on fixing them.~

Issue tracker: https://github.com/pytorch/pytorch/issues/55070

cc: ezyang bdhirsh ysiraichi 

[ghstack-poisoned]

mruberry commented:
ping @bdhirsh

bdhirsh commented:
This PR is ready, just waiting on XLA-side changes from @JackCaoG

JackCaoG commented:
Working on a XLA side fix

Lezcano commented on aten/src/ATen/native/TensorAdvancedIndexing.cpp:
```diff
@@ -173,6 +173,37 @@ TORCH_META_FUNC(scatter_add)
+    at::assert_no_overlap(result, source);
+  }
+
+  return TORCH_PRECOMPUTE_STRUCT(index_add)().set_dim(dim);
```
Just a question for me to learn. What's all this business with `TORCH_PRECOMPUTE_STRUCT` and `TORCH_PRECOMPUTE_META_FUNC`? I have not read abut this in the RFC. When should I use these?

krshrimali replied:
This was something which was added after the RFC release. To quote from the PR (https://github.com/pytorch/pytorch/pull/61746) that added this change:

> This commit introduces a new feature for structured kernels that allows
kernels to declare quantities as "precomputed" in
native_functions.yaml, compute them once in the meta function and
reuse them again in the impl.

There was a lot of redundancy, specially with usage of `maybe_wrap_dim` in both meta and impl functions, and I've mostly used this `TORCH_PRECOMPUTE_<STRUCT/META_FUNC>` to avoid redundancy (it, of course, can be used for other variables as well). These will also require some changes in the `native_functions.yaml` file.

If you don't have a quantity being computed in the meta function, and re-used in the impl function, then this is not required. I hope this helps answer your question.

lezcano reviewed (commented):
Left a few comments, all of them very minor.

krshrimali commented on aten/src/ATen/native/TensorAdvancedIndexing.cpp:
```diff
@@ -707,48 +738,34 @@ Tensor index_copy(const Tensor & self, int64_t dim, const Tensor & index, const
     AT_DISPATCH_INDEX_TYPES(index.scalar_type(), "index_add_cpu_", [&] () {
       auto index_data = index_contig.data_ptr<index_t>();
-      for (const auto i : c10::irange(numel)) {
+      for (auto i = 0; i < numel; i++) {
```
Great question, looks like it happened because of the merge with upstream after https://github.com/pytorch/pytorch/pull/66747 was merged. So basically, this is how it was before, and I'll add `c10::irange` usage again. Thanks, @Lezcano for taking a look. :)

krshrimali replied:
Fixed now. :)

Lezcano commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -9870,6 +9870,8 @@ def ref_pairwise_distance(input1, input2):
            gradcheck_nondet_tol=GRADCHECK_NONDET_TOL),
     OpInfo('index_add',
```
Actually I'm not sure now. Are the in-place variants tested by default if they exist?

krshrimali replied:
> Actually I'm not sure now. Are the in-place variants tested by default if they exist?

Thanks for the question, @Lezcano. Yes, they are tested. `inplace_variant` is set to `op_name + "_"` by default: https://github.com/pytorch/pytorch/blob/1adeeabdc0c8832420c091c5c668843768530d7f/torch/testing/_internal/common_methods_invocations.py#L645-L647

These are then tested in `test_variant_consistency_eager` (https://github.com/pytorch/pytorch/blob/master/test/test_ops.py#L452). Do you think this answers your question?

Lezcano replied:
ooops, sorry for the noise then.

mruberry commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -5671,19 +5671,21 @@
-
-- func: index_add_.alpha(Tensor(a!) self, int dim, Tensor index, Tensor source, *, Scalar alpha) -> Tensor(a!)
-  variants: method
+- func: index_add.out(Tensor self, int dim, Tensor index, Tensor source, *, Scalar alpha=1, Tensor(a!) out) -> Tensor(a!)
```
It would be nice to be consistent about having out= coverage, so that seems like a reasonable thing for one of us to say

bdhirsh replied:
Sorry for the churn @krshrimali - in that case, can you add some python docs to the out= variant? (and remove the entry from this list).

The XLA-side PR is ready, so after that I'll get ready to land :)

krshrimali replied:
So I'm a little confused for the doc fix (on adding the details for `out=` variant).

`Tensor.index_add` should not have `out=` variant, as it's not supported, right? (I tried for a few other ops as well, which support `out=` variant but they don't support for method variants).

The details on `out=` coverage should ideally go in `torch.index_add` documentation, which doesn't exist. The solution I'm thinking of is:

1. Copy the information from [in-place variant doc](https://pytorch.org/docs/stable/generated/torch.Tensor.index_add_.html?highlight=index_add#torch.Tensor.index_add_), and create an entry for `torch.index_add` in `torch/_torch_docs.py`.
2. Add `out=None` and details of that to `torch.index_add` doc (just how it's done for other ops).

And of course, remove `index_add.out` entry from `_SKIP_PYTHON_BINDINGS` list.

Please let me know what you think. cc: @bdhirsh @mruberry @Lezcano

Lezcano replied:
You are right in that methods do not have an out variant. The more you know.

So, what I had in mind was:
- ~~Modifying the signature in https://pytorch.org/docs/stable/generated/torch.Tensor.index_add.html?highlight=index_add to be `Tensor.index_add(dim, index, tensor2, *, out=None) → Tensor`~~ No need to do this as methods do not have an out variant (which I just learnt).
- Add an entry for `torch.index_add` similar to that one from `Tensor.index_add`, that is, a very very short text referencing `index_add_` and just adding the `out=` part int he signature.

krshrimali replied:
Thanks, @Lezcano - so basically just something like https://pytorch.org/docs/stable/generated/torch.Tensor.index_add.html?highlight=index_add for `torch.index_add`. Makes sense, will do that. :) Thank you for your help.

krshrimali replied:
Update: should be fixed now. Waiting for the docs to build, will recheck and request final (🤞🏻 ) review :)

krshrimali committed:
Update on "Porting index_add to structured kernels, add an out variant"


This PR attempts to port `index_add` to structured kernels, but does more than that:

* Adds an `out=` variant to `index_add`
* Revises `native_functions.yaml` registrations, to not have multiple entries and instead pass default value to `alpha`.
* Changes in `derivatives.yaml` file for autograd functioning
* Revises error messages, please see: https://github.com/pytorch/pytorch/pull/65993#issuecomment-945441615

Follow-up PRs in near future will attempt to refactor the OpInfo test, and will give another look at tests in `test/test_torch.py` for this function. (hence the use of ghstack for this)

~This is WIP because there are tests failing for `Dimname` variant on mobile/android builds, and I'm working on fixing them.~

Issue tracker: https://github.com/pytorch/pytorch/issues/55070

cc: ezyang bdhirsh ysiraichi 

[ghstack-poisoned]

krshrimali committed:
Update on "Porting index_add to structured kernels, add an out variant"


This PR attempts to port `index_add` to structured kernels, but does more than that:

* Adds an `out=` variant to `index_add`
* Revises `native_functions.yaml` registrations, to not have multiple entries and instead pass default value to `alpha`.
* Changes in `derivatives.yaml` file for autograd functioning
* Revises error messages, please see: https://github.com/pytorch/pytorch/pull/65993#issuecomment-945441615

Follow-up PRs in near future will attempt to refactor the OpInfo test, and will give another look at tests in `test/test_torch.py` for this function. (hence the use of ghstack for this)

~This is WIP because there are tests failing for `Dimname` variant on mobile/android builds, and I'm working on fixing them.~

Issue tracker: https://github.com/pytorch/pytorch/issues/55070

cc: ezyang bdhirsh ysiraichi 

[ghstack-poisoned]

bdhirsh commented:
I see one last meta test failure. Maybe you can just remove this line? https://github.com/pytorch/pytorch/blob/d176c82bd5487af25689eceba537fa41511db007/test/test_torch.py#L6733

krshrimali committed:
Update on "Porting index_add to structured kernels, add an out variant"


This PR attempts to port `index_add` to structured kernels, but does more than that:

* Adds an `out=` variant to `index_add`
* Revises `native_functions.yaml` registrations, to not have multiple entries and instead pass default value to `alpha`.
* Changes in `derivatives.yaml` file for autograd functioning
* Revises error messages, please see: https://github.com/pytorch/pytorch/pull/65993#issuecomment-945441615

Follow-up PRs in near future will attempt to refactor the OpInfo test, and will give another look at tests in `test/test_torch.py` for this function. (hence the use of ghstack for this)

~This is WIP because there are tests failing for `Dimname` variant on mobile/android builds, and I'm working on fixing them.~

Issue tracker: https://github.com/pytorch/pytorch/issues/55070

cc: ezyang bdhirsh ysiraichi 

[ghstack-poisoned]

krshrimali commented:
> I see one last meta test failure. Maybe you can just remove this line?
> 
> https://github.com/pytorch/pytorch/blob/d176c82bd5487af25689eceba537fa41511db007/test/test_torch.py#L6733

Thanks for taking a look again, @bdhirsh - sorry, I was on sick leave for a day. I've pushed the commit now, and I'll try to check back in after a few hours to see all the pending and in-progress checks pass. The current failures don't look alarming to me but please correct me if I'm wrong.

Thanks again for your help.

bdhirsh commented:
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32646426).

bdhirsh commented:
still working on landing this, there are some internal test failures

bdhirsh commented on aten/src/ATen/native/TensorAdvancedIndexing.cpp:
```diff
@@ -173,6 +173,33 @@ TORCH_META_FUNC(scatter_add)
   scatter_meta_impl(*this, self, dim, index, src, "add");
 }
 
+TORCH_PRECOMPUTE_META_FUNC(index_add)
+(const Tensor& self, int64_t dim, const Tensor& index, const Tensor& source, const Scalar& alpha) {
```
I just realized before landing this - that check that TensorIterator is handling only runs in the impl function (because that's where we construct the TensorIterator). That means that the meta function won't catch everything:

```
>>> a = torch.ones(5, 5, 5, 5)
>>> b = torch.ones(5, 5, 8, 5)
>>> i = torch.ones(5, dtype=torch.int)
# This fails properly
>>> torch.index_add(a, 0, i, b)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: The size of tensor a (5) must match the size of tensor b (8) at non-singleton dimension 1

>>> a = torch.ones(5, 5, 5, 5, device='meta')
>>> b = torch.ones(5, 5, 8, 5, device='meta')
>>> i = torch.ones(5, dtype=torch.int, device='meta')
# But this passes?
>>> torch.index_add(a, 0, i, b)
tensor(..., device='meta', size=(5, 5, 5, 5))
```

So... we need some way to run those TensorIterator checks in the meta function? That looks difficult because the inputs to `borrowing_binary_op` aren't the original inputs: they're slices of the inputs. Some options that I can think of:
- (1) try to manually emulate the checks that TI runs in this constructor. We run the risk of missing some steps that TI does though. Also, it was mentioned in this thread that we hit some weird vmap errors originally trying to do this.
- (2) try to get `index_add` to inherit from `TensorIteratorBase`. This... might work, but might be more involved / you'd have to try it out. We'd need to create the intermediates that you pass into `build_borrowing_binary_op` directly in the meta function.
- (3) Throw something like "if device == kMeta, `TensorIteratorBase::borrowing_binary_op`"  into the meta function.

What do you guys think? (2) Sounds maybe nicer, but I'm not sure if you'll run into any problems trying to do it. (3) sounds hackier but easier to add in, to correct the meta function.

krshrimali replied:
@bdhirsh - Thanks for pointing this out. Sorry for missing this earlier.

2nd option does sound better and less of a hack, but I temporarily went for the 3rd method. It does fix things, and I added a comment there for reference. I'm running a little tight on other PRs, hence couldn't try the 2nd method - but I'll give it a try this week to see if it works.

I'm sorry for the delay, but I'll try getting back on the 2nd option by the end of this week, but if you think it's okay to go for the 3rd option (and the current state of this PR) - that's also okay to me, I'll try the 2nd option and if that fixes without any failures/problems - I can raise a separate PR.

bdhirsh replied:
I'll try to get this landed and we can leave that as a followup :) LGTM

ezyang replied:
BTW the hack here is incorrect when the input is scalar

```
IndexError: select() cannot be applied to a 0-dim tensor. 
```

krshrimali committed:
Update on "Porting index_add to structured kernels, add an out variant"


This PR attempts to port `index_add` to structured kernels, but does more than that:

* Adds an `out=` variant to `index_add`
* Revises `native_functions.yaml` registrations, to not have multiple entries and instead pass default value to `alpha`.
* Changes in `derivatives.yaml` file for autograd functioning
* Revises error messages, please see: https://github.com/pytorch/pytorch/pull/65993#issuecomment-945441615

Follow-up PRs in near future will attempt to refactor the OpInfo test, and will give another look at tests in `test/test_torch.py` for this function. (hence the use of ghstack for this)

~This is WIP because there are tests failing for `Dimname` variant on mobile/android builds, and I'm working on fixing them.~

Issue tracker: https://github.com/pytorch/pytorch/issues/55070

cc: ezyang bdhirsh ysiraichi

Differential Revision: [D32646426](https://our.internmc.facebook.com/intern/diff/D32646426)

[ghstack-poisoned]

krshrimali committed:
Update on "Porting index_add to structured kernels, add an out variant"


This PR attempts to port `index_add` to structured kernels, but does more than that:

* Adds an `out=` variant to `index_add`
* Revises `native_functions.yaml` registrations, to not have multiple entries and instead pass default value to `alpha`.
* Changes in `derivatives.yaml` file for autograd functioning
* Revises error messages, please see: https://github.com/pytorch/pytorch/pull/65993#issuecomment-945441615

Follow-up PRs in near future will attempt to refactor the OpInfo test, and will give another look at tests in `test/test_torch.py` for this function. (hence the use of ghstack for this)

~This is WIP because there are tests failing for `Dimname` variant on mobile/android builds, and I'm working on fixing them.~

Issue tracker: https://github.com/pytorch/pytorch/issues/55070

cc: ezyang bdhirsh ysiraichi

Differential Revision: [D32646426](https://our.internmc.facebook.com/intern/diff/D32646426)

[ghstack-poisoned]

bdhirsh commented:
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32646426).

bdhirsh commented on aten/src/ATen/native/TensorAdvancedIndexing.cpp:
```diff
@@ -196,6 +196,49 @@ TORCH_META_FUNC(_scatter_reduce)
+    auto sourceSlice = source.select(dim, 0);
+    auto self_stride_bytes = result.stride(dim) * elementSize(result.scalar_type());
+    auto source_stride_bytes = source.stride(dim) * elementSize(source.scalar_type());
+    auto self_dim_size = result.size(dim);
```
internal builds are failing from these 3 unused variables

krshrimali replied:
My bad, this should be fixed now.

krshrimali committed:
Update on "Porting index_add to structured kernels, add an out variant"


This PR attempts to port `index_add` to structured kernels, but does more than that:

* Adds an `out=` variant to `index_add`
* Revises `native_functions.yaml` registrations, to not have multiple entries and instead pass default value to `alpha`.
* Changes in `derivatives.yaml` file for autograd functioning
* Revises error messages, please see: https://github.com/pytorch/pytorch/pull/65993#issuecomment-945441615

Follow-up PRs in near future will attempt to refactor the OpInfo test, and will give another look at tests in `test/test_torch.py` for this function. (hence the use of ghstack for this)

~This is WIP because there are tests failing for `Dimname` variant on mobile/android builds, and I'm working on fixing them.~

Issue tracker: https://github.com/pytorch/pytorch/issues/55070

cc: ezyang bdhirsh ysiraichi

Differential Revision: [D32646426](https://our.internmc.facebook.com/intern/diff/D32646426)

[ghstack-poisoned]

bdhirsh commented:
Do you mind rebasing? Looks like master was very red at the time of the last rebase

krshrimali committed:
Update on "Porting index_add to structured kernels, add an out variant"


This PR attempts to port `index_add` to structured kernels, but does more than that:

* Adds an `out=` variant to `index_add`
* Revises `native_functions.yaml` registrations, to not have multiple entries and instead pass default value to `alpha`.
* Changes in `derivatives.yaml` file for autograd functioning
* Revises error messages, please see: https://github.com/pytorch/pytorch/pull/65993#issuecomment-945441615

Follow-up PRs in near future will attempt to refactor the OpInfo test, and will give another look at tests in `test/test_torch.py` for this function. (hence the use of ghstack for this)

~This is WIP because there are tests failing for `Dimname` variant on mobile/android builds, and I'm working on fixing them.~

Issue tracker: https://github.com/pytorch/pytorch/issues/55070

cc: ezyang bdhirsh ysiraichi

Differential Revision: [D32646426](https://our.internmc.facebook.com/intern/diff/D32646426)

[ghstack-poisoned]

krshrimali commented:
> Do you mind rebasing? Looks like master was very red at the time of the last rebase

Sure, @bdhirsh - should be done now :)

bdhirsh commented:
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32646426).
